bots/twitterbot.go

```
@@ -0,0 +1,140 @@
+)
+
+type TwitterBot struct {
+	client        *twitter.Client
+	stream        *twitter.Stream
+	coniksAddress string
+	handle        string
+}
+
+var _ Bot = (*TwitterBot)(nil)
+
+type TwitterConfig struct {
+	CONIKSAddress string `toml:"coniks_address"`
+	TwitterOAuth  `toml:"twitter_oauth"`
+	Handle        string `toml:"twitter_bot_handle"`
+}
+
+type TwitterOAuth struct {
+	ConsumerKey    string
+	ConsumerSecret string
+	AccessToken    string
+	AccessSecret   string
+}
+
+func NewTwitterBot(path string) (Bot, error) {
+	var conf TwitterConfig
+	if _, err := toml.DecodeFile(path, &conf); err != nil {
+		return nil, fmt.Errorf("Failed to load config: %v", err)
+	}
+
+	auth := conf.TwitterOAuth
+	config := oauth1.NewConfig(auth.ConsumerKey, auth.ConsumerSecret)
+	token := oauth1.NewToken(auth.AccessToken, auth.AccessSecret)
+	// OAuth1 http.Client will automatically authorize Requests
+	httpClient := config.Client(oauth1.NoContext, token)
+
+	// Twitter Client
+	client := twitter.NewClient(httpClient)
+
+	bot := new(TwitterBot)
+	bot.client = client
+	bot.coniksAddress = conf.CONIKSAddress
+	bot.handle = conf.Handle
+
+	return bot, nil
+}
+
+func (bot *TwitterBot) Run() {
+	demux := twitter.NewSwitchDemux()
+	demux.DM = func(dm *twitter.DirectMessage) {
+		if strings.EqualFold(dm.SenderScreenName, bot.handle) {
+			return
+		}
+		// check if received DM has proper format
+		if strings.HasPrefix(dm.Text, "?CONIKS?") {
+			msg := strings.TrimPrefix(dm.Text, "?CONIKS?")
+			res := bot.HandleRegistration(dm.SenderScreenName, []byte(msg))
+			// Hackity, hack, hack!
+			// Twitter APIs probably doesn't want people call them so fast
+			time.Sleep(5 * time.Second)
+			err := bot.SendDM(dm.SenderScreenName, "?CONIKS?"+res)
+			if err != nil {
+				log.Printf("[registration bot] " + err.Error())
+			}
+		}
+	}
+
+	userParams := &twitter.StreamUserParams{
+		StallWarnings: twitter.Bool(true),
+	}
+	stream, err := bot.client.Streams.User(userParams)
+	if err != nil {
```

I'm surprised this doesn't complain if my access token is bogus.  Is there's any way to fail early if your credentials are no good?